<a href="https://colab.research.google.com/github/cristiandarioortegayubro/UNI/blob/main/colab/dc_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![](https://github.com/cristiandarioortegayubro/UNI/blob/main/UNI%2039%20Aniversario.jpg?raw=true)

# **Limpieza de Datos**

Nos encontramos iniciando un nuevo proyecto, nuestra empresa es una pequeña inmobiliaria que quiere fusionar el potencial del manejo de datos con la venta/alquiler de propiedades. Al ser una pequeña empresa el ***departamento de datos*** esta compuesto por una persona: nosotros. 

Esto quiere decir que vamos a absorber los roles de:

- Ingeniero de Datos, y 
- Analista de Datos.


## **Ingesta de datos de propiedades**

***El dataset central de este proyecto de propiedades de Argentina, tiene la siguientes columnas:***

- **id** - Identificador del aviso. No es único: si el aviso es actualizado por la inmobiliaria (nueva versión del aviso) se crea un nuevo registro con la misma id pero distintas fechas: de alta y de baja.
- **ad_type** - Tipo de aviso (Propiedad, Desarrollo/Proyecto).
- **start_date** - Fecha de alta del aviso.
- **end_date** - Fecha de baja del aviso.
- **created_on** - Fecha de alta de la primera versión del aviso.
- **lat** - Latitud.
- **lon** - Longitud.
- **l1** - Nivel administrativo 1: país.
- **l2** - Nivel administrativo 2: usualmente provincia.
- **l3** - Nivel administrativo 3: usualmente ciudad.
- **l4** - Nivel administrativo 4: usualmente barrio.
- **l5** - sin informacion
- **l6** - sin informacion
- **rooms** - Cantidad de ambientes
- **bedrooms** - Cantidad de dormitorios
- **bathrooms** - Cantidad de baños.
- **surface_total** - Superficie total en m².
- **surface_covered** - Superficie cubierta en m².
- **price** - Precio publicado en el anuncio.
- **currency** - Moneda del precio publicado.
- **price_period** - Periodo del precio (Diario, Semanal, Mensual)
- **title** - Título del anuncio.
- **description** - Descripción del anuncio.
- **property_type**: - Caracteristicas de la propiedad
- **operation_type**: - Caracteristicas de la operacion ['Venta', 'Alquiler', 'Alquiler temporal']

## **Habilitando bibliotecas**

***Habilitar las bibliotecas necesarias para generar un dataframe, hacer limpieza de datos, y generar gráficos.***

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


### Para analisis de datos

In [ ]:
import numpy as np
import pandas as pd

## **Análisis Exploratorio de los Datos**

***Leer el archivo  ```.csv``` con Pandas y guardarlo en una variable llamada ```prop```***

In [ ]:
prop = pd.read_csv("/content/drive/MyDrive/Notebooks/Workshop/propiedades.csv")

***Observar los primeros 8 registros***

In [ ]:
prop.head(8)

id    ad_type  start_date    end_date  created_on  \
0  DyVXfkpKygVBKuUk5olH+A==  Propiedad  2020-08-22  2020-09-03  2020-08-22   
1  9naojilaMecJN4jlQiTkGg==  Propiedad  2020-08-22  2020-09-04  2020-08-22   
2  tlCPRJPjoDEUzuuCelemAQ==  Propiedad  2020-08-22  2020-08-31  2020-08-22   
3  Zw3b91glQUO3HNrM5fPYlQ==  Propiedad  2020-08-22  2020-09-04  2020-08-22   
4  bsU81gm9JEgtZCbTYgvykg==  Propiedad  2020-08-22  2020-09-04  2020-08-22   
5  3Cips/GEPuJqktYycSALCQ==  Propiedad  2020-08-22  2020-09-04  2020-08-22   
6  6TO5dWwmivuJ67zkztUzCA==  Propiedad  2020-08-22  2020-09-04  2020-08-22   
7  VlE71iojfIFoH4piJfFQ/g==  Propiedad  2020-08-22  2020-09-04  2020-08-22   

         lat        lon         l1                            l2  \
0 -34.407468 -58.957367  Argentina      Bs.As. G.B.A. Zona Norte   
1 -37.996039 -57.542509  Argentina  Buenos Aires Costa Atlántica   
2 -31.380187 -58.009182  Argentina                    Entre Ríos   
3 -27.494106 -55.123455  Argentina                      Misiones   
4 -32.948856 -60.630464  Argentina                      Santa Fe   
5 -32.884278 -60.710901  Argentina                      Santa Fe   
6 -34.604195 -58.559827  Argentina      Bs.As. G.B.A. Zona Oeste   
7 -32.948834 -60.630169  Argentina                      Santa Fe   

                l3  ... bathrooms surface_total  surface_covered  price  \
0            Pilar  ...       NaN      133139.0              NaN    NaN   
1    Mar del Plata  ...       NaN         687.0            687.0    NaN   
2        Concordia  ...       1.0          80.0             80.0    NaN   
3            Oberá  ...       1.0           NaN              NaN    NaN   
4          Rosario  ...       1.0          76.0             66.0    NaN   
5          Rosario  ...       1.0          39.0             33.0    NaN   
6  Tres de Febrero  ...       1.0           NaN             42.0    NaN   
7          Rosario  ...       2.0         130.0            108.0    NaN   

   currency  price_period                                              title  \
0       NaN       Mensual  VENTA - Lote Industrial 130.000m2 - Parque Ind...   
1       NaN       Mensual                               Casa - Santa Cecilia   
2       NaN       Mensual                                   Casa - Concordia   
3       NaN       Mensual                                    Terreno - Obera   
4       NaN       Mensual                             Departamento - Rosario   
5       NaN       Mensual                               9 DE JULIO 535 06 01   
6       NaN       Mensual                             Departamento - Caseros   
7       NaN       Mensual                             Departamento - Rosario   

                                         description  property_type  \
0  Lote (Nro.50) mide 133.139 m2\n<br>Valor: U$S ...           Lote   
1  Conjunto edilicio de casa y departamentos en b...           Otro   
2  Casa de un dormitorio, cocina, baño, living co...           Casa   
3  IMPORTANTE PROPIEDAD EN VENTA EN OBERA MISIONE...           Lote   
4  PORTAL DE LOS MARINOS<br>Es un complejo de viv...   Departamento   
5  CARACTERÍSTICAS<br>Ingreso living comedor con ...   Departamento   
6  Departamento 2 ambientes con Lavadero en Nicar...   Departamento   
7  PORTAL DE LOS MARINOS<br>Es un complejo de viv...   Departamento   

  operation_type  
0          Venta  
1          Venta  
2       Alquiler  
3          Venta  
4          Venta  
5       Alquiler  
6       Alquiler  
7          Venta  

[8 rows x 25 columns]

***Observar los últimos 5 registros***

In [ ]:
prop.tail()

id    ad_type  start_date    end_date  \
999995  jMYhwydKqUje8HCx7RzwJQ==  Propiedad  2020-12-04  9999-12-31   
999996  FsP7ornzqnK1CvUL8quQ9Q==  Propiedad  2020-12-04  9999-12-31   
999997  fJLVUIziPl31ACoaE/fj1g==  Propiedad  2020-12-04  9999-12-31   
999998  7mW3bUSJFA0RBOj2xtBi+w==  Propiedad  2020-12-04  9999-12-31   
999999  rEynGyGrrsxNeG0ib5vUBA==  Propiedad  2020-12-04  2020-12-21   

        created_on        lat        lon         l1                        l2  \
999995  2020-12-04 -32.958863 -60.665633  Argentina                  Santa Fe   
999996  2020-12-04 -34.596587 -58.376893  Argentina           Capital Federal   
999997  2020-12-04 -34.576445 -58.432603  Argentina           Capital Federal   
999998  2020-12-04 -32.866636 -68.841144  Argentina                   Mendoza   
999999  2020-12-04 -34.447524 -58.631552  Argentina  Bs.As. G.B.A. Zona Norte   

             l3  ... bathrooms surface_total  surface_covered     price  \
999995  Rosario  ...       7.0         400.0            300.0  400000.0   
999996   Retiro  ...       5.0         465.0            465.0    3700.0   
999997  Palermo  ...       2.0         615.0            425.0  570000.0   
999998      NaN  ...      20.0         450.0            450.0  800000.0   
999999    Tigre  ...       4.0         350.0            300.0    7000.0   

        currency  price_period  \
999995       USD       Mensual   
999996       USD       Mensual   
999997       USD       Mensual   
999998       USD       Mensual   
999999       USD       Mensual   

                                                    title  \
999995  Se vende imponente casa a minutos del centro d...   
999996                        Edificio Comercial - Retiro   
999997  Edificio Comercial  en excelente zona de Paler...   
999998                       Edificio Comercial - Capital   
999999  INCREÍBLE ECASA A LA LAGUNA DESDE DIC a FEBRER...   

                                              description    property_type  \
999995  Esta excelente propiedad se encuentra ubicada ...             Casa   
999996  GRAN EDIFICIO PETIT HOTEL ESTILO FRANCÉS  CON ...  Local comercial   
999997  Edificio Comercial CON ENTRADA INDEPENDIENTE S...  Local comercial   
999998  Local 450m² con Cloaca en 25 de mayo, Argentin...  Local comercial   
999999  ESPECTACULAR CASA A LA LAGUNA  -BARRIO SANTA B...             Casa   

           operation_type  
999995              Venta  
999996           Alquiler  
999997              Venta  
999998              Venta  
999999  Alquiler temporal  

[5 rows x 25 columns]

***Imprimir la cantidad de filas y columnas del dataframe***

In [ ]:
print(f"El conjunto de datos tiene {prop.shape[0]} filas y {prop.shape[1]} columnas")

El conjunto de datos tiene 1000000 filas y 25 columnas


**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $1000000$ filas y $25$ columnas...***

**Usar info**

In [ ]:
prop.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 25 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   id               1000000 non-null  object 
 1   ad_type          1000000 non-null  object 
 2   start_date       1000000 non-null  object 
 3   end_date         1000000 non-null  object 
 4   created_on       1000000 non-null  object 
 5   lat              894233 non-null   float64
 6   lon              894191 non-null   float64
 7   l1               1000000 non-null  object 
 8   l2               1000000 non-null  object 
 9   l3               965273 non-null   object 
 10  l4               306162 non-null   object 
 11  l5               5530 non-null     object 
 12  l6               0 non-null        float64
 13  rooms            714179 non-null   float64
 14  bedrooms         649933 non-null   float64
 15  bathrooms        765122 non-null   float64
 16  surface_total    47

***Usar ```describe()``` para variables cuantitativas y mostrar el resumen con dos decimales y transpuesto (transposicion)...***

In [ ]:
prop.describe().round(2).T

count       mean         std     min       25%       50%  \
lat              894233.0     -34.38        3.04  -54.84    -34.72    -34.59   
lon              894191.0     -59.49        2.80 -122.54    -58.88    -58.49   
l6                    0.0        NaN         NaN     NaN       NaN       NaN   
rooms            714179.0       2.92        1.74    1.00      2.00      3.00   
bedrooms         649933.0       2.10        2.08  -16.00      1.00      2.00   
bathrooms        765122.0       1.71        1.09    1.00      1.00      1.00   
surface_total    477831.0     513.76     4297.24 -136.00     50.00     95.00   
surface_covered  487756.0   11656.22  4443191.26 -130.00     45.00     78.00   
price            958243.0  346094.02  5713689.88    0.00  38000.00  97000.00   

                       75%           max  
lat                 -34.42  4.963000e+01  
lon                 -58.40  1.800000e+02  
l6                     NaN           NaN  
rooms                 4.00  4.000000e+01  
bedrooms              3.00  9.000000e+02  
bathrooms             2.00  2.000000e+01  
surface_total       250.00  2.000000e+05  
surface_covered     168.00  2.147484e+09  
price            210000.00  3.100000e+09

***Usar ```describe()``` para variables cualitativas y mostrar resumen transpuesto...***

In [ ]:
prop.describe(include=object).T

count   unique  \
id              1000000  1000000   
ad_type         1000000        1   
start_date      1000000      362   
end_date        1000000      450   
created_on      1000000      362   
l1              1000000        4   
l2              1000000       43   
l3               965273     1367   
l4               306162     1070   
l5                 5530       21   
currency         955491        5   
price_period     429870        3   
title            999999   548833   
description      999958   817449   
property_type   1000000       10   
operation_type  1000000        3   

                                                           top     freq  
id                                    DyVXfkpKygVBKuUk5olH+A==        1  
ad_type                                              Propiedad  1000000  
start_date                                          2020-10-02    25159  
end_date                                            9999-12-31   208960  
created_on                                          2020-10-02    25159  
l1                                                   Argentina   983115  
l2                                             Capital Federal   265125  
l3                                                     Rosario    61746  
l4                                                    Nordelta    19479  
l5                                            BarrioPortezuelo      735  
currency                                                   USD   738157  
price_period                                           Mensual   429842  
title                                               casa-venta     3338  
description     Departamento en Alquiler de 1 dorm. en Cordoba     1980  
property_type                                     Departamento   457907  
operation_type                                           Venta   750831

In [ ]:
prop.describe(include=["object"]).T

count   unique  \
id              1000000  1000000   
ad_type         1000000        1   
start_date      1000000      362   
end_date        1000000      450   
created_on      1000000      362   
l1              1000000        4   
l2              1000000       43   
l3               965273     1367   
l4               306162     1070   
l5                 5530       21   
currency         955491        5   
price_period     429870        3   
title            999999   548833   
description      999958   817449   
property_type   1000000       10   
operation_type  1000000        3   

                                                           top     freq  
id                                    DyVXfkpKygVBKuUk5olH+A==        1  
ad_type                                              Propiedad  1000000  
start_date                                          2020-10-02    25159  
end_date                                            9999-12-31   208960  
created_on                                          2020-10-02    25159  
l1                                                   Argentina   983115  
l2                                             Capital Federal   265125  
l3                                                     Rosario    61746  
l4                                                    Nordelta    19479  
l5                                            BarrioPortezuelo      735  
currency                                                   USD   738157  
price_period                                           Mensual   429842  
title                                               casa-venta     3338  
description     Departamento en Alquiler de 1 dorm. en Cordoba     1980  
property_type                                     Departamento   457907  
operation_type                                           Venta   750831

***Ver si existen valores nulos***

In [ ]:
prop.isnull().sum().sort_values(ascending=False)

l6                 1000000
l5                  994470
l4                  693838
price_period        570130
surface_total       522169
surface_covered     512244
bedrooms            350067
rooms               285821
bathrooms           234878
lon                 105809
lat                 105767
currency             44509
price                41757
l3                   34727
description             42
title                    1
property_type            0
id                       0
ad_type                  0
l2                       0
l1                       0
created_on               0
end_date                 0
start_date               0
operation_type           0
dtype: int64

In [ ]:
prop.isna().sum().sort_values(ascending=False)

l6                 1000000
l5                  994470
l4                  693838
price_period        570130
surface_total       522169
surface_covered     512244
bedrooms            350067
rooms               285821
bathrooms           234878
lon                 105809
lat                 105767
currency             44509
price                41757
l3                   34727
description             42
title                    1
property_type            0
id                       0
ad_type                  0
l2                       0
l1                       0
created_on               0
end_date                 0
start_date               0
operation_type           0
dtype: int64

***Ver si existen valores duplicados***

In [ ]:
prop.duplicated().sum()

0

**PUNTO DE REVISIÓN**

***Antes de empezar con la limpieza de los datos el dataframe tiene $1000000$ filas y $25$ columnas...***

In [ ]:
prop.shape

(1000000, 25)

## **Limpieza de los Datos**

***Ver si los nombres de las columnas tienen algun espacio oculto que nos dificulte al tratar de invocarlas, si encontras espacios ocultos, entonces se debe eliminar esos espacios ocultos.***

In [ ]:
prop.columns

Index(['id', 'ad_type', 'start_date', 'end_date', 'created_on', 'lat', 'lon',
       'l1', 'l2', 'l3', 'l4', 'l5', 'l6', 'rooms', 'bedrooms', 'bathrooms',
       'surface_total', 'surface_covered', 'price', 'currency', 'price_period',
       'title', 'description', 'property_type', 'operation_type'],
      dtype='object')

***Ver los nulos que hay en el dataframe, en forma de porcentaje para cada columna, ordenado del mayor al menor...***

In [ ]:
(prop.isnull().sum()*100 / prop.shape[0]).round(2).sort_values(ascending=False)

l6                 100.00
l5                  99.45
l4                  69.38
price_period        57.01
surface_total       52.22
surface_covered     51.22
bedrooms            35.01
rooms               28.58
bathrooms           23.49
lat                 10.58
lon                 10.58
currency             4.45
price                4.18
l3                   3.47
title                0.00
description          0.00
property_type        0.00
id                   0.00
ad_type              0.00
l2                   0.00
l1                   0.00
created_on           0.00
end_date             0.00
start_date           0.00
operation_type       0.00
dtype: float64

***Eliminar columnas que tengan mas del $30$% de nulos (nans), luego de este paso deberian quedar $18$ columnas si lo hiciste bien...***

In [ ]:
prop.dropna(axis=1, thresh=int(prop.shape[0]*0.7), inplace=True)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $1000000$ filas y $18$ columnas...***

In [ ]:
prop.shape

(1000000, 18)

***Quedarse con solo los registros que tienen un $90$% de sus celdas libres de nulos (nans), luego de este paso deberia quedarte $922772$ registros.***

In [ ]:
prop.dropna(axis=0, thresh= int(prop.shape[1]*0.9), inplace=True)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $922772$ filas y $18$ columnas...***

In [ ]:
prop.shape

(922772, 18)

***El analisis se realizará sobre el AMBA, eliminar todo lo que no sea parte del mismo. Luego de este paso deberian quedarte $617747$ registros...***

AMBA = Bs As GBA Zona Oeste, Norte y Sur + Capital Federal

In [ ]:
prop.l2.unique()

array(['Entre Ríos', 'Misiones', 'Santa Fe', 'Bs.As. G.B.A. Zona Oeste',
       'Bs.As. G.B.A. Zona Norte', 'Tucumán',
       'Buenos Aires Costa Atlántica', 'Chubut', 'Capital Federal',
       'Córdoba', 'Buenos Aires Interior', 'Bs.As. G.B.A. Zona Sur',
       'San Luis', 'Neuquén', 'Colonia', 'Río Negro', 'Montevideo',
       'Maldonado', 'Mendoza', 'Corrientes', 'Salta', 'Florida',
       'Tierra Del Fuego', 'Chaco', 'Santa Cruz', 'La Pampa', 'Catamarca',
       'Canelones', 'Santiago Del Estero', 'San Juan', 'Miami', 'Jujuy',
       'Michigan', 'São Paulo', 'La Rioja', 'Rocha', 'Santa Catarina',
       'Pennsylvania', 'Rio Grande do Norte', 'Maryland', 'Formosa',
       'Rio de Janeiro', 'California'], dtype=object)

In [ ]:
prop = prop[(prop.l2 == "Bs.As. G.B.A. Zona Oeste") |
            (prop.l2 == "Bs.As. G.B.A. Zona Norte") |
            (prop.l2 == "Bs.As. G.B.A. Zona Sur") |
            (prop.l2 == "Capital Federal")]

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $617747$ filas y $18$ columnas...***

In [ ]:
prop.shape

(617747, 18)

***Chequear la columna l1 que representa al pais, deberia existir solo un valor: Argentina. Si hay otros valores eliminarlos...***

In [ ]:
prop.l1.unique()

array(['Argentina'], dtype=object)

***Los registros que tienen el mismo id son de la misma propiedad pero actualizado, quedarse con el mas actualizado, los demas eliminarlos.***

In [ ]:
prop.id.duplicated().sum()

0

In [ ]:
prop.id.nunique()

617747

In [ ]:
prop.id.nunique() == prop.shape[0]

True

***Eliminar las columnas:***

- **id** porque ya sabemos que todos los registros son diferentes y no vamos a usar esta columna para identificarlos.
- **l1** porque solo tiene un valor: Argentina. No aporta informacion a los datos.
- **description** y **title** porque no contamos con herramientas de NLP para volverlas informacion para visualizar o modelizar.


In [ ]:
prop = prop.drop(columns=["id","l1","description","title"])

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $617747$ filas y $14$ columnas...***

In [ ]:
prop.shape

(617747, 14)

***Sospechamos que las columnas ```start_date``` y ```created_on``` en este caso tienen los mismos valores, chequearlo, si son iguales eliminar la columna ```created_on```...***

In [ ]:
(prop.loc[:,"start_date"] == prop.loc[:,"created_on"]).sum()

617747

In [ ]:
((prop.loc[:,"start_date"] == prop.loc[:,"created_on"]).sum()) == prop.shape[0]

True

In [ ]:
prop.drop(columns="created_on", inplace=True)

***Si hiciste todo bien para este momento deberias tener un dataframe de $(617747, 13)$***

***Ver los primeros registros del dataframe...***

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $617747$ filas y $13$ columnas...***

In [ ]:
prop.shape

(617747, 13)

***Vamos a analizar la columna ```ad_type```, que tipos de datos unicos tiene, si solo tiene un tipo de dato categorico eliminarla porque no aporta información....***

In [ ]:
prop.ad_type.unique()

array(['Propiedad'], dtype=object)

In [ ]:
prop.ad_type.nunique()

1

In [ ]:
prop.drop(columns="ad_type", inplace=True)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $617747$ filas y $12$ columnas...***

In [ ]:
prop.shape

(617747, 12)

### AMBA

***Vamos a analizar la columna l3, para eso primero vamos a filtrar solo los lugares que pertenecen al AMBA:***
    


In [ ]:
ambant = ['General Rodríguez',
          'La Plata',
          'Escobar',
          'Pilar',
          'San Vicente',
          'Marcos Paz',
          'Presidente Perón',
          'Cañuelas']

In [ ]:
prop = prop[~(prop.l3.isin(ambant))]

***Para este paso deberias tener 528442 rows × 12 columns...***

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $528442$ filas y $12$ columnas...***

In [ ]:
prop.shape

(528442, 12)

### **Latitud y longitud**

***Vamos a analizar la calidad de los datos en las columnas de lat y lon. Buscaremos posibles outliers o datos incorrectos, para eso calcula para ambas columnas: el promedio, el maximo y el minimo...***

In [ ]:
prop.lat.describe().round(4)

count    500380.0000
mean        -34.6013
std           0.4330
min         -53.7886
25%         -34.6457
50%         -34.5998
75%         -34.5630
max          42.4313
Name: lat, dtype: float64

In [ ]:
prop.lon.describe().round(4)

count    500341.0000
mean        -58.4906
std           0.4543
min        -122.4553
25%         -58.5510
50%         -58.4571
75%         -58.4046
max          58.7031
Name: lon, dtype: float64

***Evidentemente hay ciertos puntos que son incorrectos, sus coordenadas caen fuera del AMBA. Para eso buscamos cuales son los valores frontera para la latitud y longitud en el AMBA.***

***Estimamos un segmento de latitudes de [-35, -34] y longitudes de [-59 a -58]***

***Eliminar todo lo que este fuera de esos rangos (no eliminar valores NaN)***

In [ ]:
prop = prop[prop.lat.between(-35, -34) &
            prop.lon.between(-59, -58)]

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $499739$ filas y $12$ columnas...***

In [ ]:
prop.shape

(499739, 12)

***Volver a chequear estadisticos de lat y lon para ver si ya saque valores incorrectos...***

In [ ]:
prop.lat.describe().round(4)

count    499739.0000
mean        -34.6054
std           0.1003
min         -34.9882
25%         -34.6457
50%         -34.5998
75%         -34.5630
max         -34.0490
Name: lat, dtype: float64

In [ ]:
prop.lon.describe().round(4)

count    499739.0000
mean        -58.4836
std           0.1102
min         -58.9819
25%         -58.5505
50%         -58.4570
75%         -58.4045
max         -58.0217
Name: lon, dtype: float64

***Eliminar los NaN de la columna l3***

In [ ]:
prop.dropna(subset=["l3"], inplace = True)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $499633$ filas y $12$ columnas...***

In [ ]:
prop.shape

(499633, 12)

***Volvamos a ver el porcentaje de nulos de cada columna...***

In [ ]:
(prop.isna().sum()*100 / prop.shape[0]).round(2).sort_values(ascending=False)

rooms             19.39
bathrooms         15.34
currency           1.05
price              0.94
start_date         0.00
end_date           0.00
lat                0.00
lon                0.00
l2                 0.00
l3                 0.00
property_type      0.00
operation_type     0.00
dtype: float64

***Veamos los valores unicos de la columna ```property_type```***

In [ ]:
prop.property_type.unique()

array(['Departamento', 'Casa', 'Oficina', 'Local comercial', 'Lote',
       'Cochera', 'PH', 'Otro', 'Depósito', 'Casa de campo'], dtype=object)

***La columna se ve bien, hacemos lo mismo con ```operation_type```***

In [ ]:
prop.operation_type.unique()

array(['Alquiler', 'Alquiler temporal', 'Venta'], dtype=object)

***La columna se ve bien. Nos quedan price, currency, bathrooms y rooms.***

***Vamos a currency, veamos los valores unicos de esta columna...***

In [ ]:
prop.currency.unique()

array([nan, 'ARS', 'USD', 'COP'], dtype=object)

***Tenemos el peso argentino, el dolar, el peso colombiano y NaN. Veamos los registros con NaN.***

In [ ]:
prop[(prop.currency.isnull())]

start_date    end_date        lat        lon  \
6       2020-08-22  2020-09-04 -34.604195 -58.559827   
10      2020-08-22  2020-08-25 -34.439610 -58.629968   
649     2021-03-21  9999-12-31 -34.804675 -58.173498   
1140    2021-03-21  2021-04-23 -34.630972 -58.424762   
1228    2021-03-21  2021-03-22 -34.610553 -58.361774   
...            ...         ...        ...        ...   
997026  2020-12-04  2021-01-27 -34.648968 -58.577701   
997029  2020-12-04  2020-12-09 -34.761623 -58.207809   
997032  2020-12-04  2021-03-16 -34.919530 -58.479625   
997053  2020-12-04  2021-01-13 -34.611887 -58.359590   
997060  2020-12-04  2021-01-02 -34.644868 -58.469612   

                              l2                  l3  rooms  bathrooms  price  \
6       Bs.As. G.B.A. Zona Oeste     Tres de Febrero    2.0        1.0    NaN   
10      Bs.As. G.B.A. Zona Norte               Tigre    4.0        4.0    NaN   
649       Bs.As. G.B.A. Zona Sur         Berazategui    2.0        2.0    NaN   
1140             Capital Federal               Boedo    4.0        1.0    NaN   
1228             Capital Federal       Puerto Madero    3.0        2.0    NaN   
...                          ...                 ...    ...        ...    ...   
997026  Bs.As. G.B.A. Zona Oeste          La Matanza    4.0        2.0    NaN   
997029    Bs.As. G.B.A. Zona Sur         Berazategui   14.0        1.0    NaN   
997032    Bs.As. G.B.A. Zona Sur  Esteban Echeverría    5.0        4.0    NaN   
997053           Capital Federal       Puerto Madero    4.0        3.0    NaN   
997060           Capital Federal              Flores    9.0        3.0    NaN   

       currency property_type     operation_type  
6           NaN  Departamento           Alquiler  
10          NaN          Casa  Alquiler temporal  
649         NaN  Departamento              Venta  
1140        NaN  Departamento              Venta  
1228        NaN  Departamento              Venta  
...         ...           ...                ...  
997026      NaN          Casa              Venta  
997029      NaN          Casa              Venta  
997032      NaN          Casa              Venta  
997053      NaN  Departamento              Venta  
997060      NaN          Casa              Venta  

[5251 rows x 12 columns]

In [ ]:
prop[(prop.currency != "ARS") &
     (prop.currency != "USD") &
     (prop.currency != "COP")]

start_date    end_date        lat        lon  \
6       2020-08-22  2020-09-04 -34.604195 -58.559827   
10      2020-08-22  2020-08-25 -34.439610 -58.629968   
649     2021-03-21  9999-12-31 -34.804675 -58.173498   
1140    2021-03-21  2021-04-23 -34.630972 -58.424762   
1228    2021-03-21  2021-03-22 -34.610553 -58.361774   
...            ...         ...        ...        ...   
997026  2020-12-04  2021-01-27 -34.648968 -58.577701   
997029  2020-12-04  2020-12-09 -34.761623 -58.207809   
997032  2020-12-04  2021-03-16 -34.919530 -58.479625   
997053  2020-12-04  2021-01-13 -34.611887 -58.359590   
997060  2020-12-04  2021-01-02 -34.644868 -58.469612   

                              l2                  l3  rooms  bathrooms  price  \
6       Bs.As. G.B.A. Zona Oeste     Tres de Febrero    2.0        1.0    NaN   
10      Bs.As. G.B.A. Zona Norte               Tigre    4.0        4.0    NaN   
649       Bs.As. G.B.A. Zona Sur         Berazategui    2.0        2.0    NaN   
1140             Capital Federal               Boedo    4.0        1.0    NaN   
1228             Capital Federal       Puerto Madero    3.0        2.0    NaN   
...                          ...                 ...    ...        ...    ...   
997026  Bs.As. G.B.A. Zona Oeste          La Matanza    4.0        2.0    NaN   
997029    Bs.As. G.B.A. Zona Sur         Berazategui   14.0        1.0    NaN   
997032    Bs.As. G.B.A. Zona Sur  Esteban Echeverría    5.0        4.0    NaN   
997053           Capital Federal       Puerto Madero    4.0        3.0    NaN   
997060           Capital Federal              Flores    9.0        3.0    NaN   

       currency property_type     operation_type  
6           NaN  Departamento           Alquiler  
10          NaN          Casa  Alquiler temporal  
649         NaN  Departamento              Venta  
1140        NaN  Departamento              Venta  
1228        NaN  Departamento              Venta  
...         ...           ...                ...  
997026      NaN          Casa              Venta  
997029      NaN          Casa              Venta  
997032      NaN          Casa              Venta  
997053      NaN  Departamento              Venta  
997060      NaN          Casa              Venta  

[5251 rows x 12 columns]

***Vemos que la cantidad de NaN en esta columna es del $1$%, sin esta columna no podremos saber en que moneda esta el precio, por lo que tampoco sabremos el precio (el precio es la columna mas interesante de este dataset). Decidimos eliminar los registros con NaN de la columna currency.***

In [ ]:
prop.dropna(subset=["currency"], inplace= True)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $494382$ filas y $12$ columnas...***

In [ ]:
prop.shape

(494382, 12)

***Volver a analizar el porcentaje de NaN de cada columna.***

In [ ]:
(prop.isna().sum()*100 / prop.shape[0]).round(2).sort_values(ascending=False)

rooms             19.57
bathrooms         15.49
start_date         0.00
end_date           0.00
lat                0.00
lon                0.00
l2                 0.00
l3                 0.00
price              0.00
currency           0.00
property_type      0.00
operation_type     0.00
dtype: float64

***Vemos que se eliminaron los NaN de price al eliminar los de currency, matamos 2 pajaros de un tiro.***

Ahora si crear una nueva columna **llamada precio** con el valor de las propiedades **en dolares**. Utilizar estas cotizaciones

1 dolar = $283 pesos argentinos

1 dolar = $4637 pesos colombianos

In [ ]:
condicion = [(prop.currency == "ARS"),
             (prop.currency == "COP"),
             (prop.currency == "USD")]

In [ ]:
eleccion = [(prop.price / 283),
            (prop.price / 4637),
            (prop.price)]

In [ ]:
prop["precio"] = np.select(condicion, eleccion)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $494382$ filas y $13$ columnas...***

In [ ]:
prop.shape

(494382, 13)

### Alquiler temporal

***Vamos a centrar el analisis en Alquiler y Venta, eliminamos los registros de 'Alquiler temporal'...***

In [ ]:
prop = prop[(prop.operation_type != "Alquiler temporal")]

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $472774$ filas y $13$ columnas...***

In [ ]:
prop.shape

(472774, 13)

### Registros entre 40 dolares o 5000000

***Eliminar los registros cuyo precio sea inferior a 40 dolares o mayor a 5.000.000 dolares...***

In [ ]:
prop[(prop.precio.between(40, 5000000, inclusive="neither"))].shape

(467968, 13)

In [ ]:
prop[(prop.precio.between(40, 5000000, inclusive="neither"))].shape

(467968, 13)

In [ ]:
prop = prop[(prop.precio.between(40, 5000000, inclusive="neither"))]

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $467968$ filas y $13$ columnas...***

In [ ]:
prop.shape

(467968, 13)

***Eliminar las ventas cuyo precio sea menor a 5000 dolares***

In [ ]:
prop = prop.loc[~((prop.operation_type == "Venta") & (prop.precio < 5000)),:]

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $467586$ filas y $13$ columnas...***

In [ ]:
prop.shape

(467586, 13)

***Vamos ahora a la columna bathrooms, eliminamos los registros cuyo valor en bathrooms sea mayor al de rooms + 1***

In [ ]:
prop = prop[~(prop.bathrooms > (prop.rooms + 1))]

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $465964$ filas y $13$ columnas...***

In [ ]:
prop.shape

(465964, 13)

### Mas de 8 baños o mas de 15 habitaciones

***Eliminar los registros que tengan mas de $8$ baños o mas de $15$ habitaciones...***

In [ ]:
prop = prop[~((prop.bathrooms > 8) | (prop.rooms > 15))]

***Eliminar la columna ```price``` y ```currency```, ya que vamos a utilizar la columna precio...***

In [ ]:
prop.drop(columns=["price", "currency"], inplace=True)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $465451$ filas y $11$ columnas...***

In [ ]:
prop.shape

(465451, 11)

### ***Rellenado***

- ***Rellenar los NaN de bathrooms con la mediana de la columna bathrooms, hacer lo mismo con la columna rooms.***

Pero antes vamos a marcar los registros a los cuales vamos a inferir, para eso **primero de rellenar**, crea una columna llamada **rellenado** la cual tendra un 1 si alguno de sus campos fue rellenado por nosotros o 0 si no fue rellenado.

In [ ]:
condicion2 = [(prop.rooms.isna()),
              (prop.bathrooms.isna()),
              (~(prop.rooms.isna())),
              (~(prop.bathrooms.isna()))]

In [ ]:
seleccion2 = [1,1,0,0]

In [ ]:
prop["rellenado"] = np.select(condicion2, seleccion2)

***Calcular la cantidad de registros rellenados y no rellenados...***

In [ ]:
prop.rellenado.value_counts()

0    345611
1    119840
Name: rellenado, dtype: int64

***Ahora si: Rellená los NaN de bathrooms con la mediana de la columna bathrooms, hacer lo mismo con la columna rooms con la mediana de la columna rooms...***

In [ ]:
bathrooms_median = prop.bathrooms.median()
rooms_median = prop.rooms.median()

In [ ]:
prop.bathrooms.fillna(value=bathrooms_median, inplace=True)
prop.rooms.fillna(value=rooms_median, inplace = True)

***Volver a analizar el porcentaje de NaN de cada columna. No deberia de existir ningun NaN en nuestro dataframe...***

In [ ]:
(prop.isna().sum()*100/prop.shape[0])

start_date        0.0
end_date          0.0
lat               0.0
lon               0.0
l2                0.0
l3                0.0
rooms             0.0
bathrooms         0.0
property_type     0.0
operation_type    0.0
precio            0.0
rellenado         0.0
dtype: float64

***Parece ser que el formato se encuentra correcto, vamos a aplicar una pequeña comprobacion logica: la fecha start_date puede ser igual o menor que end_date, nunca mayor. Si ocurre esto, eliminar los registros que contienen estos errores.***

In [ ]:
prop[(prop.start_date <= prop.end_date)]

start_date    end_date        lat        lon  \
16      2020-08-22  2020-09-04 -34.633053 -58.419906   
17      2020-08-22  2020-09-04 -34.397401 -58.649185   
18      2020-08-22  2020-09-04 -34.397401 -58.649185   
19      2020-08-22  2020-09-04 -34.598112 -58.400784   
20      2020-08-22  2020-09-04 -34.604054 -58.393770   
...            ...         ...        ...        ...   
999990  2020-12-04  2020-12-16 -34.462664 -58.625235   
999991  2020-12-04  2021-01-05 -34.473210 -58.760048   
999993  2020-12-04  9999-12-31 -34.482248 -58.501989   
999996  2020-12-04  9999-12-31 -34.596587 -58.376893   
999997  2020-12-04  9999-12-31 -34.576445 -58.432603   

                              l2                   l3  rooms  bathrooms  \
16               Capital Federal                Boedo    3.0        1.0   
17      Bs.As. G.B.A. Zona Norte                Tigre    3.0        2.0   
18      Bs.As. G.B.A. Zona Norte                Tigre    3.0        2.0   
19               Capital Federal             Recoleta    3.0        3.0   
20               Capital Federal            Balvanera    3.0        1.0   
...                          ...                  ...    ...        ...   
999990  Bs.As. G.B.A. Zona Norte                Tigre    3.0        5.0   
999991  Bs.As. G.B.A. Zona Norte  Malvinas Argentinas    3.0        2.0   
999993  Bs.As. G.B.A. Zona Norte           San Isidro   11.0        4.0   
999996           Capital Federal               Retiro   12.0        5.0   
999997           Capital Federal              Palermo   14.0        2.0   

          property_type operation_type        precio  rellenado  
16              Oficina       Alquiler  2.473498e+02          1  
17              Oficina       Alquiler  1.236749e+02          1  
18              Oficina       Alquiler  9.540636e+01          1  
19              Oficina       Alquiler  2.296820e+02          1  
20              Oficina       Alquiler  2.120141e+02          1  
...                 ...            ...           ...        ...  
999990             Casa          Venta  1.200000e+06          1  
999991             Casa          Venta  1.120000e+05          1  
999993             Casa          Venta  6.900000e+05          0  
999996  Local comercial       Alquiler  3.700000e+03          0  
999997  Local comercial          Venta  5.700000e+05          0  

[465451 rows x 12 columns]

***Vamos a pasar las columnas start_date y end_date, que son strings, a formato datetime. Recorda que el formato de fecha que tenemos en nuestras columnas de fecha es %Y/%m/%d***

In [ ]:
prop.start_date = pd.to_datetime(prop["start_date"], format = "%Y/%m/%d")

In [ ]:
prop.end_date = pd.to_datetime(prop["end_date"], format = "%Y/%m/%d")

OutOfBoundsDatetime: ignored

***Al tratar de pasar la columna end_date a datetime nos sale un error, hay un registro que lo esta causando, investigar que registro causa este error...***

***Hay muchos registros que tienen como end_date 9999-12-31, como la documentacion es de mala calidad, la misma no especifica que significa que el end_date tenga este valor, pero aplicando un poco de sentido comun deducimos que los registros que tienen esta fecha significa que el aviso no se dio de baja, por lo cual no se vendio.***

***Cambiar los registros con esta fecha, cambiandola a la fecha de la primera publicacion de python por parte de Guido van Rossum:***
**1991-02-20**. 

***Los registros con esta fecha seran los que aun no se vendieron.***

In [ ]:
prop.end_date.replace("9999-12-31","1991-02-20", inplace=True)

***Ahora si pasar la columna end_date a datetime...***

In [ ]:
prop.end_date = pd.to_datetime(prop["end_date"], format = "%Y/%m/%d")

***Vamos a cambiar el nombre de las columnas de la siguiente manera:...***

- start_date => inicio
- end_date => fin
- lat => lat
- lon => lon
- l2 => zona
- l3 => municipio
- rooms => ambientes
- bathrooms => baños
- property_type => tipo
- operation_type => operacion
- precio => precio
- rellenado => rellenado

In [ ]:
nombres = ["inicio", "fin", "lat", "lon", "zona", "municipio", "ambientes", "baños", "tipo", "operacion", "precio", "rellenado" ]

In [ ]:
prop.columns = nombres

In [ ]:
prop.columns

Index(['inicio', 'fin', 'lat', 'lon', 'zona', 'municipio', 'ambientes',
       'baños', 'tipo', 'operacion', 'precio', 'rellenado'],
      dtype='object')

***Crear dos columnas nuevas:***

- ***año_inicio:*** tiene que contener el numero de año que hay en la fecha de la columna inicio
- ***mes_inicio:*** tiene que contener el numero de mes que hay en la fecha de la columna inicio

In [ ]:
prop["año_inicio"] = pd.DatetimeIndex(prop.inicio).year
prop["mes_inicio"] = pd.DatetimeIndex(prop.fin).month

***Por ultimo, reinicia el indice...***

In [ ]:
prop.reset_index(drop=True, inplace=True)

### ***Comprobacion***

***Si hasta aca lo hiciste bien te deberia quedar un dataset de $465451$ rows $×$ $14$ columns y que tenga los siguientes parametros iguales:...***



In [ ]:
if (465451 - prop.shape[0]) == 0:
  print(f"Vamo nomaaaaaaaaaa")
else:
  print(f"Segui participando, todavia tenes una diferencia de {465451 - prop.shape[0]}")

Vamo nomaaaaaaaaaa
